keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,360 @@
+  ObeservedGeneration int64
+  // CompletedGeneration is the greatest spec generation whose **migration worker**
+  // has completed.
+  CompletedGeneration int64
+}
+```
+
+[#67428]:https://github.com/kubernetes/kubernetes/issues/67428
+[continue token]:https://github.com/kubernetes/kubernetes/blob/972e1549776955456d9808b619d136ee95ebb388/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L82
+[workflow section]:#alpha-workflow
+
+We need to expand the discovery API to recognize APIs in different groups
+that share the same storage, e.g., `extensions/v1beta1/deployment` and
+`apps/v1/deployment` share the same key space in etcd. Specifically, we expand
+the [APIResource][] API:
+
+```golang
+type APIResource struct {
+  Name string
+  ...
+  // StorageUID=SHA-256(etcd server in the form of scheme://ip:port, etcd path),
+  // obfuscated with a random key. E.g., for pods, it is
+  // SHA-256(scheme://ip:port/registry/pods).
+  // The string is obfuscated to avoid exposing server configuration details.
+  StorageUID string
+}
+```
+
+[APIResource]:https://github.com/kubernetes/kubernetes/blob/d38fb2c6451447ad0ecbdfed971a563419d60883/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L874
+
+
+### Failure recovery
+
+As stated in the goals section, the migration has to make progress even if the
+environment is flaky. This section describes how the migrator recovers from
+failure.
+
+Kubernetes **replicaset controller** restarts the **migration controller** `pod`
+if it fails. The restarted **migration controller** follows the same bootstrap
+workflow as described in the [workflow section][]. Because the **migration
+controller** adopts existing `migration` objects, it can resume from where
+it left off.
+
+Kubernetes **job controller** ensures the completion of a **migration worker**.
+The **migration worker** uses the continue token as the checkpoint.
+* the worker stores the latest continue token returned by the apiserver in a
+  `configmap`.
+* when the worker starts, it resumes listing with the continue token.
+* the **migration controller** creates the configmap when creates the worker `job`,
+  and sets the worker job as the owner. When the **migration controller** deletes
+  the completed `job`, the `configmap` is garbage collected.
+
+[workflow section]:#alpha-workflow
+
+### Beta workflow - Automation
+
+It is a beta goal to automate the migration workflow. That is, migration does
+not need to be triggered manually by cluster admins, or by custom control loops
+of Kubernetes distributions.
+
+The automated migration should work for Kubernetes built-in resource types,
+custom resources, and aggregated resources.
+
+Here are two alternatives and their requirements.
+
+#### a. Exposing the configured storage versions via API
+
+This is what the [Active Management][] design proposes. It requires:
+* exposing the configured [storage versions][] of Kubernetes built-in resource
+  types and aggregated resources. 
```

Comment:
I still think this is the way to go. If an aggregated apiserver doesn't want to participate in the storage migrator or doesn't actually store its data or doesn't support ordinary changes or whatever, it can just omit the information in the discovery API.